### PR TITLE
vstart: specify the NVME PCI BUS ID when enable spdk.

### PR DESCRIPTION
**Pull request by majianpeng**
Currently we use vendor::device to specify the nvme name when using
spdk. But it have some problems
a)it can't use different vendor or different device with the same
device.
b)if machine has same vendor&device, we don't choose specific device.

So we diretly specify the PCI BUS ID for NVME device.

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
@tchaikov. please review.

File: src/vstart.sh [hunk 1]
```diff
@@ -221,7 +221,7 @@ usage=$usage"\t--short: short object names only; necessary for ext4 dev\n"
 usage=$usage"\t--nolockdep disable lockdep\n"
 usage=$usage"\t--multimds <count> allow multimds with maximum active count\n"
 usage=$usage"\t--without-dashboard: do not run using mgr dashboard\n"
-usage=$usage"\t--bluestore-spdk <vendor>:<device>: enable SPDK and specify the PCI-ID of the NVME device\n"
+usage=$usage"\t--bluestore-spdk: enable SPDK and specify the PCI-ID(etc, 0000:81:00.0) of the NVME device with comma-separated\n"
```
Comment (tchaikov) [inline src/vstart.sh:224]: you are removing the argument spec. but we still need it for specifying the PCI-ID, right?

> with comma-separated

could you fix the syntax error? i think what you mean is "a comma-delimited list of PCI-IDs of NVME device.

File: src/vstart.sh [hunk 2]
```diff
@@ -510,14 +510,6 @@ wconf() {
...
 }
 
-get_pci_selector() {
-    which_pci=$1
-    lspci -mm -n -D -d $pci_id | cut -d ' ' -f 1 | sed -n $which_pci'p'
-}
-
-get_pci_selector_num() {
-    lspci -mm -n -D -d $pci_id | cut -d' ' -f 1 | wc -l
-}
```
Comment (tchaikov) [inline src/vstart.sh:520]: why do you think by removing these helpers we can make it easier to choose a PCI device?

File: src/vstart.sh [hunk 3]
```diff
@@ -632,14 +624,6 @@ EOF
...
     if [ "$objectstore" == "bluestore" ]; then
         if [ "$spdk_enabled" -eq 1 ]; then
-            if [ "$(get_pci_selector_num)" -eq 0 ]; then
-                echo "Not find the specified NVME device, please check." >&2
-                exit
-            fi
-            if [ $(get_pci_selector_num) -lt $CEPH_NUM_OSD ]; then
-                echo "OSD number ($CEPH_NUM_OSD) is greater than NVME SSD number ($(get_pci_selector_num)), please check." >&2
-                exit
-            fi
```
Comment (tchaikov) [inline src/vstart.sh:642]: i still these safe guards are nice to have.

File: src/vstart.sh (hunk 2 above)
Comment (majianpeng) [reply src/vstart.sh:520]: Those helper  make vendor::device to pci bus .  i specify these parameters directly.

**Comment by majianpeng** [comment]
the original parameter is vendor::device. If my machine have different vendor NVME or same vendor but different type device. I only use one type.

File: src/vstart.sh (hunk 2 above)
Comment (tchaikov) [reply src/vstart.sh:520]: @majianpeng my point is that this change does not make the life easier if one *have to* specify them directly.

Comment (majianpeng) [reply src/vstart.sh:520]: @tchaikov .  Like parameter --bluestore-devs, we set the device name directly.  Before vstart.sh, we should call spdk/scritps/setup.sh. we can get all infos of nvme, like:

>>0000:81:00.0 (8086 2701): Skipping un-whitelisted NVMe controller nvme0n1
>>0000:b5:00.0 (8086 0a54): nvme -> uio_pci_generic
>>0000:b6:00.0 (8086 0a54): nvme -> uio_pci_generic
>>0000:b7:00.0 (8086 0a54): nvme -> uio_pci_generic
>>0000:b8:00.0 (8086 0a54): nvme -> uio_pci_generic
From that, we know all nvme pci bus info .

**Comment by majianpeng** [comment]
@tchaikov  only update command description by your suggestion..

**Comment by yison** [comment]
@tchaikov Hi, this patch is helpful to me. I can specify some NVME devices by using PCI Bus ID when using vstart. For example: MON=1 OSD=2 RGW=0 MDS=0 MGR=1 ../src/vstart.sh  -d -n -x --bluestore-spdk 0000:18:00.0,0000:da:00.0 --no-parallel

**Comment by tchaikov** [comment]
> @tchaikov Hi, this patch is helpful to me. I can specify some NVME devices by using PCI Bus ID when using vstart. For example: MON=1 OSD=2 RGW=0 MDS=0 MGR=1 ../src/vstart.sh -d -n -x --bluestore-spdk 0000:18:00.0,0000:da:00.0 --no-parallel

@yison @majianpeng will take a look at this PR late this week.

File: src/vstart.sh [hunk 4]
```diff
@@ -222,7 +221,7 @@ usage=$usage"\t--short: short object names only; necessary for ext4 dev\n"
 usage=$usage"\t--nolockdep disable lockdep\n"
 usage=$usage"\t--multimds <count> allow multimds with maximum active count\n"
 usage=$usage"\t--without-dashboard: do not run using mgr dashboard\n"
-usage=$usage"\t--bluestore-spdk <vendor>:<device>: enable SPDK and specify the PCI-ID of the NVME device\n"
+usage=$usage"\t--bluestore-spdk: enable SPDK and with a comma-delimited list of PCI-IDs of NVME device(etc,0000:81:00.0)\n"
```
Comment (tchaikov) [inline src/vstart.sh:224]:
```suggestion
usage=$usage"\t--bluestore-spdk: enable SPDK and with a comma-delimited list of PCI-IDs of NVME device (e.g, 0000:81:00.0)\n"
```

**Pull request merged**